File: apps/parent-mobile/modules/play-integrity/android/src/main/java/expo/modules/digihostelplayintegrity/DigihostelPlayIntegrityModule.kt
package expo.modules.digihostelplayintegrity

import com.google.android.gms.tasks.Tasks
import com.google.android.play.core.integrity.IntegrityManagerFactory
import com.google.android.play.core.integrity.IntegrityTokenRequest
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * ADR-003 implementation — thin wrapper around Google's official Play
 * Integrity Standard API (`com.google.android.play:integrity`), NOT a
 * hand-rolled attestation mechanism. This module's only job is to obtain a
 * genuine, Google-signed integrity token bound to a server-issued nonce and
 * hand the opaque token back to JS — it never inspects, decodes, or makes any
 * trust decision about the token itself. The actual verdict decision happens
 * exclusively server-side (apps/api/src/domain/device/attestationVerifier.ts)
 * — this class could return a token for an entirely fraudulent request and
 * the backend would still correctly reject it; this module is a courier, not
 * an authority.
 *
 * Requires: a real device (or an emulator running a "Google Play" system
 * image, which the AVD used in Task 8's own physical-device pivot did NOT
 * have) with Google Play services installed and up to date. Real physical-
 * device verification of this exact call is the ADR-003 implementation
 * report's own tracked next step, gated on Google Play Console/Cloud Project
 * provisioning — see that report for the full status.
 */
class DigihostelPlayIntegrityModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("DigihostelPlayIntegrity")

    // AsyncFunction rather than Function: requestIntegrityToken() returns a
    // Play Services Task, an inherently async network+device operation, and
    // must never block the JS thread.
    AsyncFunction("requestIntegrityToken") { nonce: String, cloudProjectNumber: String ->
      val context = appContext.reactContext
        ?: throw CodedException("ERR_NO_CONTEXT", "No Android context available", null)

      val integrityManager = IntegrityManagerFactory.create(context)
      val request = IntegrityTokenRequest.builder()
        .setNonce(nonce)
        .setCloudProjectNumber(cloudProjectNumber.toLong())
        .build()

      try {
        // AsyncFunction already runs off the JS thread; Tasks.await() here
        // blocks only this background thread, not the UI/JS thread, which is
        // the documented, correct way to bridge a Play Services Task into a
        // synchronous-looking call inside an Expo AsyncFunction lambda.
        val response = Tasks.await(integrityManager.requestIntegrityToken(request))
        response.token()
      } catch (e: Exception) {
        // Deliberately surfaced as a real error, never swallowed into a
        // fabricated "success" token — see this file's own doc comment and
        // ADR-003 implementation task §11 (every failure must fail closed).
        throw CodedException(
          "ERR_PLAY_INTEGRITY_REQUEST_FAILED",
          "Play Integrity token request failed: ${e.message}",
          e,
        )
      }
    }
  }
}

File: apps/parent-mobile/modules/play-integrity/src/DigihostelPlayIntegrityModule.ts
import { NativeModule, requireNativeModule } from "expo";

declare class DigihostelPlayIntegrityModule extends NativeModule<Record<string, never>> {
  /**
   * ADR-003 implementation — requests a genuine Google Play Integrity token
   * bound to `nonce` (the exact value returned by the backend's
   * POST /devices/challenge — see src/services/devices/devices.ts). Never
   * decodes or interprets the resulting token itself; it is relayed as-is to
   * POST /devices/register, where the real trust decision is made
   * server-side. Rejects (never resolves with a fabricated value) if Google
   * Play services is unavailable/out of date or the request otherwise fails.
   */
  requestIntegrityToken(nonce: string, cloudProjectNumber: string): Promise<string>;
}

// This call loads the native module object from the JSI. Android-only — see
// expo-module.config.json; there is no iOS/apple native backing for this
// module (App Attest is a separate, not-yet-implemented integration, see the
// ADR-003 implementation report).
export default requireNativeModule<DigihostelPlayIntegrityModule>("DigihostelPlayIntegrity");
